File: camaleonte/camaleonte/python-cc/src/mediums/google_api.py
import io
import os
import time

from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.http import (
    MediaFileUpload,
    MediaIoBaseUpload,
    MediaIoBaseDownload,
)

from googleapiclient.errors import HttpError
import time

SCOPES = ['https://www.googleapis.com/auth/drive']
TOKEN_PATH = "creds/token.json"


class GoogleDriveAPI:
    """A wrapper for Google Drive API v3."""

    def __init__(self):
        self.service_worker = None

    def authenticate_drive(self, credentials_path: str) -> None:
        """
        Authenticates with the Google Drive API using OAuth 2.0.
        Requires a client_secrets.json at credentials_path.
        """
        creds = None
        if not os.path.isfile(TOKEN_PATH):
            flow = InstalledAppFlow.from_client_secrets_file(credentials_path, SCOPES)
            creds = flow.run_local_server(port=0)
            with open(TOKEN_PATH, 'w') as token_file:
                token_file.write(creds.to_json())

        creds = Credentials.from_authorized_user_file(TOKEN_PATH, SCOPES)
        self.service_worker = build('drive', 'v3', credentials=creds)

    def upload_file_to_drive(self, file_path: str, destination_id: str) -> None:
        """
        Uploads a file from disk to Google Drive.

        Args:
            file_path: Local path to the file.
            destination_id: Drive folder ID to upload into.
        """
        assert self.service_worker, "Authenticate first."
        assert os.path.exists(file_path), f"No such file: {file_path}"
        metadata = {'name': os.path.basename(file_path), 'parents': [destination_id]}
        media = MediaFileUpload(file_path, resumable=True)
        self.service_worker.files().create(
            body=metadata, media_body=media, fields='id,name'
        ).execute()

    def download_file_from_drive_bytes(self, target_id: str) -> bytes:
        """
        Downloads a Drive file into memory and returns its bytes.

        Args:
            target_id: ID of the Drive file.
        """
        assert self.service_worker, "Authenticate first."
        request = self.service_worker.files().get_media(fileId=target_id)
        buffer = io.BytesIO()
        downloader = MediaIoBaseDownload(buffer, request)
        done = False
        while not done:
            status, done = downloader.next_chunk()
        buffer.seek(0)
        return buffer.read()

    def download_file_from_drive(self, destination: str, target_id: str) -> None:
        """
        Downloads a Drive file to local disk.

        Args:
            destination: Local file path to write.
            target_id: ID of the Drive file.
        """
        data = self.download_file_from_drive_bytes(target_id)
        with open(destination, 'wb') as f:
            f.write(data)

    def get_file_properties(self, target_id: str) -> dict:
        """
        Retrieves appProperties for a single file.

        Args:
            target_id: Drive file ID.
        """
        assert self.service_worker, "Authenticate first."
        resp = self.service_worker.files().get(
            fileId=target_id, fields='appProperties'
        ).execute()
        return resp.get('appProperties', {})

    def update_properties(self, file_id: str, properties: dict) -> dict:
        """
        Updates appProperties of a single file.

        Args:
            file_id: Drive file ID.
            properties: dict of key→value to set.
        Returns:
            The updated file resource.
        """
        assert self.service_worker, "Authenticate first."
        body = {'appProperties': properties}
        return self.service_worker.files().update(
            fileId=file_id, body=body
        ).execute()

    def list_files(self,
                   directory_id: str = None,
                   filename: str = None,
                   ignore_directories: bool = False) -> list[dict]:
        """
        Lists files in a Drive folder, optionally filtering.

        Args:
            directory_id: Folder ID to search within.
            filename: Exact filename to filter.
            ignore_directories: If True, skip folders.
        """
        assert self.service_worker, "Authenticate first."
        all_files = []
        page_token = None
        parts = []
        if directory_id:
            parts.append(f"'{directory_id}' in parents")
        if filename:
            parts.append(f"name = '{filename}'")
        if ignore_directories:
            parts.append("mimeType != 'application/vnd.google-apps.folder'")
        query = " and trashed=false and ".join(parts) if parts else "trashed=false"

        while True:
            resp = self.service_worker.files().list(
                q=query,
                orderBy="name",
                pageSize=1000,
                pageToken=page_token,
                fields="nextPageToken, files(id,name,mimeType,parents)"
            ).execute()
            all_files.extend(resp.get('files', []))
            page_token = resp.get('nextPageToken')
            if not page_token:
                break

        return all_files

    def watch_file(self, target_id: str, poll_interval=0.1, timeout=300) -> bool:
        """
        Polls a file until modified or trashed.

        Args:
            target_id: Drive file ID.
            poll_interval: Seconds between polls.
            timeout: Max seconds to wait.
        """
        assert self.service_worker, "Authenticate first."
        start = time.time()
        meta = self.service_worker.files().get(
            fileId=target_id, fields='modifiedTime,trashed'
        ).execute()
        last_mod = meta['modifiedTime']

        while True:
            if time.time() - start > timeout:
                return False
            time.sleep(poll_interval)
            meta = self.service_worker.files().get(
                fileId=target_id, fields='modifiedTime,trashed'
            ).execute()
            if meta['trashed'] or meta['modifiedTime'] != last_mod:
                return True

    def edit_file(self, target_id: str, new_file_path: str) -> None:
        """
        Replaces file content by uploading a new file.

        Args:
            target_id: Drive file ID.
            new_file_path: Local path to replacement file.
        """
        assert self.service_worker, "Authenticate first."
        media = MediaFileUpload(new_file_path, resumable=True)
        self.service_worker.files().update(
            fileId=target_id, media_body=media
        ).execute()

    def edit_file_bytes(self, target_id: str, bytes_: bytes) -> None:
        """
        Replaces file content with in-memory bytes.

        Args:
            target_id: Drive file ID.
            bytes_: Byte data or BytesIO.
        """
        assert self.service_worker, "Authenticate first."
        buffer = bytes_ if isinstance(bytes_, io.BytesIO) else io.BytesIO(bytes_)
        buffer.seek(0)
        media = MediaIoBaseUpload(buffer, mimetype='application/octet-stream', resumable=False)
        self.service_worker.files().update(
            fileId=target_id, media_body=media
        ).execute()

    def delete_file(self, target_id: str) -> None:
        """Moves a file to Trash."""
        assert self.service_worker, "Authenticate first."
        self.service_worker.files().delete(fileId=target_id).execute()

    def restore_file(self, target_id: str) -> None:
        """Restores a trashed file."""
        assert self.service_worker, "Authenticate first."
        self.service_worker.files().update(
            fileId=target_id, body={'trashed': False}
        ).execute()

    def empty_bin(self) -> None:
        """Permanently deletes all trashed files."""
        assert self.service_worker, "Authenticate first."
        self.service_worker.files().emptyTrash().execute()

    def clear_all_file_properties_in_folder(self, folder_id: str) -> int:
        """
        Clears appProperties on every file in a folder.

        Args:
            folder_id: Drive folder ID.
        Returns:
            Number of files cleared.
        """
        assert self.service_worker, "Authenticate first."
        count = 0
        page_token = None

        while True:
            resp = self.service_worker.files().list(
                q=f"'{folder_id}' in parents and trashed=false",
                fields="nextPageToken, files(id,name,appProperties)",
                pageSize=1000,
                pageToken=page_token
            ).execute()

            for f in resp.get('files', []):
                props = f.get('appProperties', {})
                if props:
                    self.service_worker.files().update(
                        fileId=f['id'],
                        body={'appProperties': {k: None for k in props}}
                    ).execute()
                    count += 1

            page_token = resp.get('nextPageToken')
            if not page_token:
                break

        return count

    def update_properties_batch(self,
                                file_properties_map: dict,
                                batch_size: int = 500,
                                max_retries: int = 3,
                                delay: float = 1.0) -> None:
        """
        Updates up to batch_size files per batch request, with simple retries
        on 5xx (transient) errors and a fixed delay between attempts.
        """
        items = list(file_properties_map.items())

        for start in range(0, len(items), batch_size):
            sub_map = dict(items[start:start + batch_size])

            for attempt in range(1, max_retries + 1):
                batch = self.service_worker.new_batch_http_request()

                def _cb(request_id, response, exception):
                    if exception:
                        print(f"[ERROR] {request_id}: {exception}")

                # Queue up each file update
                for fid, props in sub_map.items():
                    req = self.service_worker.files().update(
                        fileId=fid,
                        body={'appProperties': props}
                    )
                    batch.add(req, callback=_cb, request_id=fid)

                try:
                    batch.execute()
                    # on success, break out of retry loop
                    break

                except HttpError as e:
                    status = getattr(e.resp, 'status', None)
                    # only retry on server/transient errors
                    if status and status >= 500:
                        print(f"[RETRY] batch #{start//batch_size+1} attempt {attempt} failed (status={status}), retrying in {delay}s…")
                        time.sleep(delay)
                        continue
                    # non‑transient (e.g. 403 propertyCountLimitExceeded) → re‑raise
                    raise

            else:
                # exhausted retries
                print(f"[FAIL] batch #{start//batch_size+1} failed after {max_retries} attempts.")
               
    def get_properties_batch(self, file_ids: list[str]) -> dict:
        """
        Retrieves appProperties for multiple files in one batch.

        Args:
            file_ids: List of Drive file IDs.
        Returns:
            { file_id: appProperties dict, … }
        """
        assert self.service_worker, "Authenticate first."
        out = {}

        batch = self.service_worker.new_batch_http_request()

        def _cb(req_id, resp, exc):
            if not exc:
                out[resp['id']] = resp.get('appProperties', {})

        for fid in file_ids:
            req = self.service_worker.files().get(
                fileId=fid, fields='id,appProperties'
            )
            batch.add(req, callback=_cb)

        batch.execute()
        return out

File: camaleonte/camaleonte/metasploit/camaleonte/util/mediums/google_api.rb
require 'googleauth'
require 'googleauth/stores/file_token_store'
require 'google/apis/drive_v3'
require 'stringio'
require 'time'

class GoogleDriveAPI
  SCOPES = [Google::Apis::DriveV3::AUTH_DRIVE]

  def initialize
    @service_worker = Google::Apis::DriveV3::DriveService.new
  end

  # --- Authentication ---
  def authenticate_drive(credentials_path, token_path, user_id = 'default')
    client_id = Google::Auth::ClientId.from_file(credentials_path)
    token_store = Google::Auth::Stores::FileTokenStore.new(file: token_path)
    authorizer = Google::Auth::UserAuthorizer.new(client_id, SCOPES, token_store)
    credentials = authorizer.get_credentials(user_id)

    if credentials.nil?
      url = authorizer.get_authorization_url(base_url: 'urn:ietf:wg:oauth:2.0:oob')
      puts "Open this URL in your browser and enter the code:\n#{url}"
      code = gets.chomp
      credentials = authorizer.get_and_store_credentials_from_code(
        user_id: user_id, code: code, base_url: 'urn:ietf:wg:oauth:2.0:oob'
      )
    end

    @service_worker.authorization = credentials
  end

  # --- File Upload/Download ---
  def upload_file_to_drive(file_path, destination_id)
    raise 'Authenticate first.' unless @service_worker
    raise "No such file: #{file_path}" unless File.exist?(file_path)

    metadata = Google::Apis::DriveV3::File.new(
      name: File.basename(file_path),
      parents: [destination_id]
    )

    @service_worker.create_file(
      metadata,
      upload_source: file_path,
      content_type: 'application/octet-stream'
    )
  end

  def download_file_from_drive_bytes(target_id)
    raise 'Authenticate first.' unless @service_worker

    buffer = StringIO.new
    @service_worker.get_file(target_id, download_dest: buffer)
    buffer.rewind
    buffer.read
  end

  def download_file_from_drive(destination, target_id)
    data = download_file_from_drive_bytes(target_id)
    File.open(destination, 'wb') { |f| f.write(data) }
  end

  # --- Metadata ---
  def get_file_properties(target_id)
    raise 'Authenticate first.' unless @service_worker

    file = @service_worker.get_file(target_id, fields: 'appProperties')
    file.app_properties || {}
  end

  def update_properties(file_id, properties)
    raise 'Authenticate first.' unless @service_worker

    file = Google::Apis::DriveV3::File.new(app_properties: properties)
    @service_worker.update_file(file_id, file, fields: 'id,appProperties')
  end

  def update_file_properties(file_id, properties)
    update_properties(file_id, properties)
  end

  def update_properties_batch(file_properties_map, batch_size: 500, max_retries: 3, delay: 1.0)
    raise 'Authenticate first.' unless @service_worker

    items = file_properties_map.to_a

    items.each_slice(batch_size).with_index(1) do |batch_items, batch_index|
      tries = 0
      begin
        tries += 1
        @service_worker.batch do |batch|
          batch_items.each do |file_id, props|
            file = Google::Apis::DriveV3::File.new(app_properties: props)
            batch.add(
              @service_worker.update_file(file_id, file, fields: 'id,appProperties'),
              request_id: file_id
            ) do |res, err|
              # puts "[ERROR] #{file_id}: #{err}" if err
            end
          end
        end
      rescue Google::Apis::ServerError => e
        if tries < max_retries
          # puts "[RETRY] Batch ##{batch_index} attempt #{tries} failed, retrying in #{delay}s..."
          sleep delay
          retry
        else
          # puts "[FAIL] Batch ##{batch_index} failed after #{max_retries} attempts."
        end
      end
    end
  end

  def get_properties_batch(file_ids)
    raise 'Authenticate first.' unless @service_worker

    out = {}
    @service_worker.batch do |batch|
      file_ids.each do |fid|
        batch.add(@service_worker.get_file(fid, fields: 'id,appProperties'), request_id: fid) do |file, err|
          out[file.id] = file.app_properties || {} if file && err.nil?
        end
      end
    end
    out
  end

  def clear_all_file_properties_in_folder(folder_id)
    raise 'Authenticate first.' unless @service_worker

    count = 0
    page_token = nil
    loop do
      resp = @service_worker.list_files(
        q: "'#{folder_id}' in parents and trashed=false",
        fields: 'nextPageToken, files(id,name,appProperties)',
        page_size: 1000,
        page_token: page_token
      )
      resp.files.each do |file|
        props = file.app_properties || {}
        unless props.empty?
          update_properties(file.id, {}) # clear all by overwriting
          count += 1
        end
      end
      page_token = resp.next_page_token
      break if page_token.nil?
    end
    count
  end

  # --- Listing and Watching ---
  def list_files(directory_id: nil, filename: nil, ignore_directories: false)
    raise 'Authenticate first.' unless @service_worker

    query_parts = ['trashed = false']
    query_parts << "'#{directory_id}' in parents" if directory_id
    query_parts << "name = '#{filename}'" if filename
    query_parts << "mimeType != 'application/vnd.google-apps.folder'" if ignore_directories
    query = query_parts.join(' and ')

    files = []
    page_token = nil
    loop do
      resp = @service_worker.list_files(
        q: query,
        order_by: 'name',
        page_size: 1000,
        fields: 'nextPageToken, files(id,name,mimeType,parents)',
        page_token: page_token
      )
      files.concat(resp.files)
      page_token = resp.next_page_token
      break if page_token.nil?
    end
    files
  end

  def watch_file(target_id, poll_interval = 0.1, timeout = 300)
    raise 'Authenticate first.' unless @service_worker

    start_time = Time.now
    meta = @service_worker.get_file(target_id, fields: 'modifiedTime,trashed')
    last_mod = meta.modified_time

    loop do
      return false if Time.now - start_time > timeout

      sleep(poll_interval)
      meta = @service_worker.get_file(target_id, fields: 'modifiedTime,trashed')
      return true if meta.trashed || meta.modified_time != last_mod
    end
  end

  # --- Edit Files ---
  def edit_file(target_id, new_file_path)
    raise 'Authenticate first.' unless @service_worker

    res = @service_worker.update_file(
      target_id,
      nil,
      upload_source: new_file_path,
      content_type: 'application/octet-stream',
      fields: 'id'
    )
  end

  def edit_file_bytes(target_id, bytes_)
    raise 'Authenticate first.' unless @service_worker

    buffer = bytes_.is_a?(StringIO) ? bytes_ : StringIO.new(bytes_)
    res = @service_worker.update_file(
      target_id,
      nil,
      upload_source: buffer,
      content_type: 'application/octet-stream',
      fields: 'id'
    )
  end

  # --- Delete/Restore/Trash ---
  def delete_file(target_id)
    raise 'Authenticate first.' unless @service_worker

    @service_worker.delete_file(target_id)
  end

  def restore_file(target_id)
    raise 'Authenticate first.' unless @service_worker

    file = Google::Apis::DriveV3::File.new(trashed: false)
    @service_worker.update_file(target_id, file, fields: 'id')
  end

  def empty_bin
    raise 'Authenticate first.' unless @service_worker

    @service_worker.empty_trash
  end
end
